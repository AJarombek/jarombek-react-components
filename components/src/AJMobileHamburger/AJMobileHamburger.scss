/**
 * Stylesheet for the AJMobileHamburger component.
 * Author: Andrew Jarombek
 * Date: 11/24/2019
 */

.aj-mobile-hamburger {
  display: inline-block;
  cursor: pointer;
  height: 3px;
  width: 25px;
  padding: 10px 2px;

  button {
    display: block;
    padding: 0;
    border: none;
    background-color: transparent;

    &:focus {
      outline: none;
    }

    span, span:before, span:after {
      height: 3px;
      width: 25px;
      border-radius: 1px;
      display: block;
      cursor: pointer;
      background-color: black;
      position: absolute;
      content: "";
      transition: all 50ms ease;
    }

    .aj-mobile-hamburger-active {
      background-color: transparent;
    }

    span:before {
      top: -8px;
    }

    span.aj-mobile-hamburger-active:before {
      top: 0;
      transform: rotate(45deg);
    }

    span:after {
      top: 8px;
    }

    span.aj-mobile-hamburger-active:after {
      top: 0;
      transform: rotate(-45deg);
    }
  }
}
