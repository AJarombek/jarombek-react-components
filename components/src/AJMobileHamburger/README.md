### Overview

Reusable mobile hamburger component used for mobile headers on websites.

**Use cases:**

```javascript
import { AJMobileHamburger } from 'jarombek-react-components';

const CustomComponent = () => 
  <AJMobileHamburger />
```

### Files

| Filename                   | Description                                                                               |
|----------------------------|-------------------------------------------------------------------------------------------|
| `AJMobileHamburger.js`     | The reusable `AJMobileHamburger` component.                                               |
| `index.js`                 | Index file which exports the `AJMobileHamburger` component.                               |
| `styles.js`                | JSS styles for the `AJMobileHamburger` component.                                         |
