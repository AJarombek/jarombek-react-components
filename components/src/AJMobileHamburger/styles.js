/**
 * Stylesheet for the AJMobileHamburger component using JSS.
 * Author: Andrew Jarombek
 * Date: 12/30/2019
 */

export default {
  ajMobileHamburger: {
    display: 'inline-block',
    cursor: 'pointer',
    height: '3px',
    width: '25px',
    padding: '10px 2px',

    '& > button': {
      display: 'block',
      padding: 0,
      border: 'none',
      backgroundColor: 'transparent',

      '&:focus': {
        outline: 'none',
      },

      '& > span, & > span:before, & > span:after': {
        height: '3px',
        width: '25px',
        borderRadius: '1px',
        display: 'block',
        cursor: 'pointer',
        backgroundColor: 'black',
        position: 'absolute',
        content: '""',
        transition: 'all 50ms ease',
      },

      '& > span:before': {
        top: '-8px',
      },

      '& > span:after': {
        top: '8px',
      },
    },
  },
  ajMobileHamburgerActive: {
    backgroundColor: 'transparent !important',

    '&:before': {
      top: '0 !important',
      transform: 'rotate(45deg)',
    },

    '&:after': {
      top: '0 !important',
      transform: 'rotate(-45deg)',
    },
  },
  ajMobileHamburgerInactive: {},
};
