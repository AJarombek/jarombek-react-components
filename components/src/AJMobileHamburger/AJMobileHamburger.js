/**
 * Reusable mobile hamburger component.  Used for mobile headers on websites.
 * @author Andrew Jarombek
 * @since 11/16/2019
 */

import React, {useState, forwardRef} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import {createUseStyles} from 'react-jss';

import styles from './styles';

const useStyles = createUseStyles(styles);

/**
 * Component representing a hamburger icon which animates to an 'X' icon when clicked.  Commonly
 * used with mobile menu dropdowns.
 * @param onClick Additional behavior that occurs when the component is clicked.
 * @param ref A ref forwarded through the component to the <div> DOM element.
 * @return {*} React elements that represent a hamburger icon.
 */
const AJMobileHamburger = forwardRef(({onClick}, ref) => {
  const classes = useStyles();
  const [active, setActive] = useState(false);

  const spanClassName = active ?
    classnames('aj-mobile-hamburger-active', classes.ajMobileHamburgerActive):
    classnames('aj-mobile-hamburger-inactive', classes.ajMobileHamburgerInactive);

  const hamburgerClasses = classnames('aj-mobile-hamburger', classes.ajMobileHamburger);

  const onClickAction = () => {
    setActive(!active);
    if (onClick) {
      onClick();
    }
  };

  return (
    <div className={hamburgerClasses} onClick={onClickAction} ref={ref}>
      <button type="button">
        <span className={spanClassName} />
      </button>
    </div>
  );
});

AJMobileHamburger.propTypes = {
  onClick: PropTypes.func
};

export default AJMobileHamburger;
