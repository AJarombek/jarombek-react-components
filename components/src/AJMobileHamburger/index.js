import AJMobileHamburger from './AJMobileHamburger';

export default AJMobileHamburger;
