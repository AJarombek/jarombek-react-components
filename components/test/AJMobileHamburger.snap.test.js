import React from 'react';
import AJMobileHamburger from '../src/AJMobileHamburger/AJMobileHamburger';
import renderer from 'react-test-renderer';

/**
 * Snapshot test for the {@link ../../components/AJMobileHamburger} component.
 * @author Andrew Jarombek
 * @since 12/19/2019
 */

it('renders correctly', () => {
  const tree = renderer.create(<AJMobileHamburger />).toJSON();
  expect(tree).toMatchSnapshot();
});
