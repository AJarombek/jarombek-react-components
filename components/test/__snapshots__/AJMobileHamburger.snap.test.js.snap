// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="aj-mobile-hamburger ajMobileHamburger-0-2-1"
  onClick={[Function]}
>
  <button
    type="button"
  >
    <span
      className="aj-mobile-hamburger-inactive ajMobileHamburgerInactive-0-2-3"
    />
  </button>
</div>
`;
